### PR TITLE
[WIP] Peertube package (opening for collaboration+help)

### PR DESCRIPTION
**Pull request by mohe2015**
###### Motivation for this change
@Izorkin I think I fixed the import from derivation problem that is not allowed in nixpkgs.
`nix --option allow-import-from-derivation false build -L .#peertube`

(Edit: Removed outdated information)

https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/javascript.section.md are some docs that helped.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Izorkin** [comment]
@mohe2015 to use nodejs v16_x needed add `nodejs = pkgs.nodejs-16_x;`
```
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 69c104b89aa..eba82840809 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -26874,7 +26874,9 @@ with pkgs;

   peek = callPackage ../applications/video/peek { };

-  peertube = callPackage ../servers/peertube { };
+  peertube = callPackage ../servers/peertube {
+    nodejs = pkgs.nodejs-16_x;
+  };

   pflask = callPackage ../os-specific/linux/pflask {};
```

The package error building after this commit - https://github.com/Chocobozzz/PeerTube/commit/134006b0d4528c07f0022c6e75257bf9fedbfebb
The `esbuild` module is not installed correctly:
```
Generating browser application bundles (phase: building).../build/peertube-3.4.0/client/node_modules/esbuild/bin/esbuild:2
throw new Error(`esbuild: Failed to install correctly
^

Error: esbuild: Failed to install correctly

Make sure you don't have "ignore-scripts" set to true. You can check this with
"npm config get ignore-scripts". If that returns true you can reset it back to
false using "npm config set ignore-scripts false" and then reinstall esbuild.

If you're using npm v7, make sure your package-lock.json file contains either
"lockfileVersion": 1 or the code "hasInstallScript": true. If it doesn't have
either of those, then it is likely the case that a known bug in npm v7 has
corrupted your package-lock.json file. Regenerating your package-lock.json file
should fix this issue.

    at Object.<anonymous> (/build/peertube-3.4.0/client/node_modules/esbuild/bin/esbuild:2:7)
    at Module._compile (internal/modules/cjs/loader.js:1072:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1101:10)
    at Module.load (internal/modules/cjs/loader.js:937:32)
    at Function.Module._load (internal/modules/cjs/loader.js:778:12)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:76:12)
    at internal/main/run_main_module.js:17:47
```
The `esbuild` package is already in nixpkgs, but I was unable to add it - not found `esbuild` binary on build.

**Comment by mohe2015** [comment]
Let's see what ofborg has to say about this...

**Comment by mohe2015** [comment]
I assume the files are too large to be included here but as far as I know this is the only "real" possibility without import-from-derivation-like features.

**Comment by mohe2015** [comment]
Well https://github.com/NixOS/nixpkgs/pull/128749 would be a possibility

**Comment by happysalada** [comment]
What a coincidence, I was just working on this today.
I opened an issue upstream to document the fact that angular/cli is not into their package.json
https://github.com/Chocobozzz/PeerTube/issues/4393

I also had problems building on the server, I added an issue here
https://github.com/Chocobozzz/PeerTube/issues/4394

I had a couple of fixes for this that I am testing right now on this PR

File: pkgs/servers/peertube/default.nix [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    pname = "peertube";
+    inherit version;
+    src = source;
+    phases = [ "unpackPhase" "patchPhase" "installPhase" ];
+    patches = [ ./fix_yarn_lock.patch ];
+    installPhase = ''
+      mkdir $out
+      cp -a . $out/
+    '';
+  };
```
Comment (happysalada) [inline pkgs/servers/peertube/default.nix:39]:
```suggestion
```

I've just confirmed, if you use a yarn.nix file, the patching of the yarn.lock is not necessary.
tested on darwin.

**Comment by happysalada** [comment]
Just a thought, but if we could upstream the changes that were needed to make the package.json file work, then we wouldn't need to commit those lockfiles, just the yarn.nix.
Could we try to open an issue upstream with the problems in their package.json ?
(happy to do it myself if you tell me how the package.json files have been modified)

**Comment by happysalada** [comment]
I have an idea of how not to fetch the lock files and package.json to nixpkgs. We would still need to commid the yarn.nix, but at least we can divide the size by 2.

the idea is as follows.
- make a PR upstream with the updates to the package.json and yarn.lock files that you require
- use fetchpatch to get the patch resulting from the PR.

This would have 2 advantages.
- let upstream know what changes need to be made to their package.json and lock (they can accept or refuse, but at least they know)
- we don't need to commit the patched files, we can just get the source, and patch it with the patch from the PR.

Let me know what you think.

**Comment by mohe2015** [comment]
@happysalada Actually I just had messed up my original commits and importer the root files in the client derivation. As far as I know the current state in here actually works.

**Comment by mohe2015** [comment]
> Just a thought, but if we could upstream the changes that were needed to make the package.json file work, then we wouldn't need to commit those lockfiles, just the yarn.nix.
> Could we try to open an issue upstream with the problems in their package.json ?
> (happy to do it myself if you tell me how the package.json files have been modified)

They're actually unmodified I don't fully understand how yarn2nix works and other examples used all three files. But maybe then this would work without them.

**Comment by mohe2015** [comment]
The first attempt here didn't work so I retried and changed as little as possible from the original derivation. Feel free to push the fixes here but otherwise I would probably look into it in ~3 hours or so

**Comment by mohe2015** [comment]
https://github.com/NixOS/nixpkgs/pull/119110 should now contain the necessary changes.

**Pull request closed without merging**